### PR TITLE
allow quorum fileInfo to pick same parityBlocks

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
allow quorum fileInfo to pick same parityBlocks

## Motivation and Context
Bonus: allow replication to proceed without performing 
a HEAD (statObject) first, there is no value in doing 
that instead we should allow PUT to proceed and let it 
fail with pre-condition check.

## How to test this PR?
Requires a special setup, please ping if you want
to test this. 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/bucket-replication.go [hunk 1]
```diff
@@ -1349,53 +1349,6 @@ func (ri ReplicateObjectInfo) replicateAll(ctx context.Context, objectAPI Object
...
-		logger.LogIf(ctx, fmt.Errorf("unable to replicate %s/%s (%s). Target returned %s error on HEAD", bucket, object, objInfo.VersionID, cerr))
-		sendEvent(eventArgs{
-			EventName:  event.ObjectReplicationNotTracked,
-			BucketName: bucket,
-			Object:     objInfo,
-			UserAgent:  "Internal: [Replication]",
-			Host:       globalLocalNodeName,
-		})
-		return
-	}
```
Comment (poornas) [inline cmd/bucket-replication.go:1398]: if the HEAD is removed, every metadata update is a full overwrite - this is not ok. We already optimize to do a HEAD only for replication failure healing, not for inline replication

**Review by poornas — CHANGES_REQUESTED** [review]

File: cmd/bucket-replication.go [hunk 2]
```diff
@@ -1385,16 +1385,23 @@ func (ri ReplicateObjectInfo) replicateAll(ctx context.Context, objectAPI Object
...
-			BucketName: bucket,
-			Object:     objInfo,
-			UserAgent:  "Internal: [Replication]",
-			Host:       globalLocalNodeName,
-		})
-		return
+	if cerr != nil {
+		errResp := minio.ToErrorResponse(cerr)
+		switch errResp.Code {
+		case "NoSuchKey", "NoSuchVersion", "SlowDown":
```
Comment (poornas) [inline cmd/bucket-replication.go]: why try to replicate when other cluster is asking to slow down - we can just leave it for catch up?

File: cmd/bucket-replication.go [hunk 3]
```diff
@@ -1385,16 +1385,23 @@ func (ri ReplicateObjectInfo) replicateAll(ctx context.Context, objectAPI Object
...
-			BucketName: bucket,
-			Object:     objInfo,
-			UserAgent:  "Internal: [Replication]",
-			Host:       globalLocalNodeName,
-		})
-		return
+	if cerr != nil {
+		errResp := minio.ToErrorResponse(cerr)
+		switch errResp.Code {
+		case "NoSuchKey", "NoSuchVersion", "SlowDownRead":
```
Comment (harshavardhana) [inline cmd/bucket-replication.go:1391]: It must be a quorum read issue, if that is the case we are better off replicating the new state() if possible @poornas - we shouldn't ignore such situations.

Comment (harshavardhana) [reply cmd/bucket-replication.go:1391]: we should use replication as a "healing" mechanism here to make sure both sites are always in sync.

File: cmd/erasure-object.go [hunk 4]
```diff
@@ -1007,14 +1010,23 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
...
+	if opts.CheckPrecondFn != nil && !opts.NoLock {
+		ns := er.NewNSLock(bucket, object)
+		lkctx, err := ns.GetLock(ctx, globalOperationTimeout)
+		if err != nil {
+			return ObjectInfo{}, err
 		}
-		if opts.CheckPrecondFn(obj) {
+		ctx = lkctx.Context()
+		defer ns.Unlock(lkctx)
+		opts.NoLock = true
```
Comment (poornas) [inline cmd/erasure-object.go]: above will not enforce precondition if opts.NoLock is true. 

```
if opts.CheckPrecondFn != nil {
      if !opts.NoLock {
		ns := er.NewNSLock(bucket, object)
		lkctx, err := ns.GetLock(ctx, globalOperationTimeout)
		..
```

**Review by poornas — APPROVED** [review]
LGTM

**Review by vadmeste — APPROVED** [review]
LGTM but lint is failing

**Pull request merged**